src/index_notation/transformations.cpp

```diff
@@ -1487,6 +1492,310 @@ ir::Stmt GEMM::replaceValidStmt(IndexStmt stmt,
+  }
+
+  auto getIndexSpace = [&](ir::Expr reg) {
+    auto vals = ir::GetProperty::make(reg, ir::TensorProperty::Values);
+    auto logreg = ir::MethodCall::make(vals, "get_logical_region", {}, false /* deref */, Auto);
+    return ir::MethodCall::make(logreg, "get_index_space", {}, false /* deref */, Auto);
+  };
+  auto aIndexSpace = getIndexSpace(tvars[0]);
+  auto bIndexSpace = getIndexSpace(tvars[1]);
+  auto cIndexSpace = getIndexSpace(tvars[2]);
+
+  // Unpack the domains for each variable.
+  auto aDomain = ir::Var::make("aDomain", Auto);
+  auto bDomain = ir::Var::make("bDomain", Auto);
+  auto cDomain = ir::Var::make("cDomain", Auto);
+  results.push_back(ir::VarDecl::make(aDomain, ir::Call::make("runtime->get_index_space_domain", {ctx, aIndexSpace}, Auto)));
+  results.push_back(ir::VarDecl::make(bDomain, ir::Call::make("runtime->get_index_space_domain", {ctx, bIndexSpace}, Auto)));
+  results.push_back(ir::VarDecl::make(cDomain, ir::Call::make("runtime->get_index_space_domain", {ctx, cIndexSpace}, Auto)));
+
+  // Break out if either domain is empty.
+  auto bVolZero = ir::Eq::make(ir::MethodCall::make(bDomain, "get_volume", {}, false, Auto), 0);
+  auto cVolZero = ir::Eq::make(ir::MethodCall::make(cDomain, "get_volume", {}, false, Auto), 0);
+  auto guard = ir::Or::make(bVolZero, cVolZero);
+  results.push_back(ir::IfThenElse::make(guard, ir::Return::make(ir::Expr())));
+
+  const TensorVar& aTensorVar = this->content->tensorVars[0];
+  const TensorVar& bTensorVar = this->content->tensorVars[1];
+  const TensorVar& cTensorVar = this->content->tensorVars[2];
+
+  auto aAccess = ir::GetProperty::make(tvars[0], ir::TensorProperty::ValuesWriteAccessor, aTensorVar.getOrder());
+  if (inReduction) {
+    aAccess = ir::GetProperty::make(tvars[0], ir::TensorProperty::ValuesReductionAccessor, aTensorVar.getOrder());
+  }
+  auto bAccess = ir::GetProperty::make(tvars[1], ir::TensorProperty::ValuesReadAccessor, bTensorVar.getOrder());
+  auto cAccess = ir::GetProperty::make(tvars[2], ir::TensorProperty::ValuesReadAccessor, cTensorVar.getOrder());
+
+  auto getBounds = [] (ir::Expr e, std::string func) {
+    return ir::MethodCall::make(e, func, {}, false, Int());
+  };
+  auto getLen = [&] (ir::Expr domain, ir::Expr dim) {
+    return ir::Add::make(1, ir::Sub::make(ir::Load::make(getBounds(domain, "hi"), dim), ir::Load::make(getBounds(domain, "lo"), dim)));
+  };
+  auto getPtr = [&] (ir::Expr access, ir::Expr domain) {
+    return ir::MethodCall::make(access, "ptr", {getBounds(domain, "lo")}, false /* deref */, Auto);
+  };
+  auto type = Type(aTensorVar.getType().getDataType());
+  auto getStride = [&] (ir::Expr access, ir::Expr dim) {
+    return ir::Div::make(
+      ir::Load::make(ir::FieldAccess::make(ir::FieldAccess::make(access, "accessor", false, Auto), "strides", false, Int()), dim),
+      ir::Sizeof::make(type)
+    );
+  };
+
+  std::string unconst_cast;
+  {
+    // TODO(timothygu): is there a better way to do this?
```

Comment:
There's a slightly better way of doing "const_cast<" + util::toString(...) + "*>", but any template instantiations in this IR will require string manipulation unfortunately.